src/providers/mdal/qgsmdalprovider.cpp

```diff
@@ -241,6 +241,22 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
   return false;
 }
 
+void QgsMdalProvider::reloadData()
+{
+  if ( mMeshH )
+    MDAL_CloseMesh( mMeshH );
+
+  QByteArray curi = dataSourceUri().toUtf8();
+  mMeshH = MDAL_LoadMesh( curi.constData() );
+
+  if ( mMeshH )
+  {
+    const QString proj = MDAL_M_projection( mMeshH );
+    if ( !proj.isEmpty() )
+      mCrs.createFromString( proj );
+  }
```

Comment:
what happens if mesh layer has assigned some external datasets? what if they are no longer compatible with the mesh?

Reply:
may be a control to check vertices count and faces coun,. If not the same, external datasets are removed ?

Reply:
that should do the job

Reply:
After consideration about extra datasets, and sorry for this long post :

It seems the control of the extra dataset files is the responsability of the provider.

For the memory provider, maybe the reload function doesn't have sense, so, I purpose to not re-implement the reloadData() method.

For the MDAL provider, it is the MDAL API that loads and control the dataset. So when the mesh is not compatible, the behavior depends of each MDAL's drivers ... I don't have analyze the code of all the drivers, but it seems if the mesh is not compatible, the dataset is not added in the MDAL::Mesh. Can you confirm?

So, when mesh is reloaded, if a extra dataset file is no valid or the mesh is not compatible, to synchronize the Dataset stored in the MDAL::Mesh, the simple choice is to remove the uri of the file from mExtraDatasetUris list.

But I am not sure it is the best choice for the user because the extra dataset will disappear from QGIS and can't come back if the file return to be valid or compatible, except if the user reload it.

Maybe, it is more convenient for the user if the uri of the file is kept in the mExtraDatasetUris list, so its reload would be tried again during the next reload of the layer. But it is necessary to synchronize this list with the list of DatasetGroup stored in the MDAL::Mesh. To do that, I would purpose to add a void DatasetGroup in the MDAL::Mesh when QgsMdalProvider::reloadData() method is called and the extra Dataset file is not valid or incompatible. But I don't find a way to do that without modify the MDAL API... Maybe I am wrong?

As I can't implement the second solution directly in QGIS code, I retained the first one. I have added tests for this solution (with an new test data file).